File: BC - w4/Gun/Gun.cpp
#include "gun.h"

        Gun::Gun(const std::string& model, int capacity) :
        model(model),
        capacity(capacity),
        amount(0),
        totalShots(0),
        isReady(false){}
        
        Gun::~Gun(){};

        int Gun::getAmount() const{
            return this->amount;
        }
        int Gun::getCapacity() const{
            return this->capacity;
        }
        bool Gun::ready() const{
            return this->isReady;
        }
        const std::string& Gun::getModel() const{
            return this->model;
        }
        int Gun::getTotalShots() const{
            return this->totalShots;
        }

        void Gun::prepare(){
            isReady = !isReady;
        }

        void Gun::reload(){
            this->amount = capacity;
        }
        void Gun::shoot(){
            if (isReady == false) {
                throw NotReady ();
            }
            if (amount == 0) {
                throw OutOfRounds();
            }
            amount -=1;
            totalShots +=1;
            std::cout << "Bang!" <<std::endl;
        }

        std::ostream& operator<<(std::ostream& out, const Gun& gun) {
            out << "model"<<" "<<gun.getModel()<<std::endl;
            out << "capacit"<<" "<<gun.getCapacity()<<std::endl;
            out << "amount"<<" "<<gun.getAmount()<<std::endl;
            out << "ready"<<" "<<gun.ready()<<std::endl;
            out << "totalShots"<<" "<<gun.getTotalShots()<<std::endl;
            return out;
        }

File: BC - w4/java/Gun/Gun.java
package Gun;


public class Gun {
	private int amount;
	private int capacity;
	private boolean isReady;
	private String model;
	private int totalShots;
	
	
   Gun( String model, int capacity){
	   this.capacity = capacity;
	   this.model = model;
	   this.amount = 0;
	   this.totalShots = 0;
	   this.isReady = false;   
    	
    }

    int getAmount() {
        return this.amount;
    }
    int getCapacity() {
        return this.capacity;
    }
    boolean ready() {
        return this.isReady;
    }
    
    String getModel() {
        return this.model;
    }
    int getTotalShots() {
        return this.totalShots;
    }

    void prepare(){
        isReady = !isReady;
    }

    void reload(){
        this.amount = capacity;
    }
    void shoot() throws NotReady, outOfRounds {
        if (isReady == false) {
            throw new NotReady ();
        }
        if (amount == 0) {
            throw new outOfRounds();
        }
        amount -=1;
        totalShots +=1;
        System.out.println("Bang!");
    }
    
    public void print(){
		System.out.println("model "+this.model+"\n"
							+"capacity " + this.capacity+"\n"
							+"amount " + this.amount+ "\n"
							+"ready " + this.isReady+ "\n" 
							+"totalShots " +this.totalShots);
	}
}
